core/store/src/flat_state.rs

```diff
@@ -645,12 +653,39 @@ impl FlatStorageState {
         Ok(vec![])
     }
 
-    // Update the head of the flat storage, including updating the flat state in memory and on disk
-    // and updating the flat state to reflect the state at the new head
+    /// Requests updating flat head.
+    /// If new head is not a child of current flat head, does nothing and returns Ok. It can happen when we have a fork:
+    ///
+    ///      (flat head)        /-------> 6
+    /// 1 ->      2     -> 3 -> 4
+    ///                         \---> 5
+    ///
+    /// where for (5) we can call `update_flat_head(3)` and then for (6) we can call `update_flat_head(2)`.
+    /// Otherwise updates the head of the flat storage, including updating the flat state in memory and on disk
+    /// and updating the flat state to reflect the state at the new head.
     // TODO (#7327): implement garbage collection of old deltas.
     #[cfg(feature = "protocol_feature_flat_state")]
     pub fn update_flat_head(&self, new_head: &CryptoHash) -> Result<(), FlatStorageError> {
         let mut guard = self.0.write().expect(POISONED_LOCK_ERR);
+        let current_head_height = guard
+            .blocks
+            .get(&guard.flat_head)
+            .ok_or(guard.create_block_not_supported_error(&guard.flat_head))?
+            .height;
+        let new_head_height = match guard.blocks.get(new_head) {
+            Some(block_info) => block_info.height,
+            None => {
+                // If new head is not supported, it is behind the current head, so we can return Ok.
+                return Ok(());
+            }
+        };
+        // If new head height is not greater than current head height, new head is not a child of flat head, so we
+        // return Ok.
+        if new_head_height <= current_head_height {
```

Comment:
Actually I wonder if it's cleaner to not change this function and check at the its callsite to ignore the "BlockNotFoundError". A lot of the checks you did here is included at `get_deltas_between_blocks`

Reply:
Yeah, I agree. If `get_deltas_between_blocks` fails, then new code fails as well.